fingerprint/infra/scanner/src/main/java/com/simprints/fingerprint/infra/scanner/v2/scanner/ota/cypress/CypressOtaController.kt

```diff
@@ -27,55 +24,47 @@ import io.reactivex.rxkotlin.toObservable
  */
 class CypressOtaController(private val crc32Calculator: Crc32Calculator) {
 
-    private inline fun <reified R : CypressOtaResponse> sendCypressOtaModeCommandAndReceiveResponse(
-        cypressOtaMessageChannel: CypressOtaMessageChannel,
-        errorHandler: ResponseErrorHandler,
-        command: CypressOtaCommand
-    ): Single<R> =
-        cypressOtaMessageChannel.sendCypressOtaModeCommandAndReceiveResponse<R>(command)
-            .handleErrorsWith(errorHandler)
 
-    fun program(
-        cypressOtaMessageChannel: CypressOtaMessageChannel,
-        errorHandler: ResponseErrorHandler,
-        firmwareBinFile: ByteArray
-    ): Observable<Float> =
-        sendPrepareDownloadCommand(cypressOtaMessageChannel, errorHandler)
-            .andThen(
-                sendDownloadCommand(cypressOtaMessageChannel, errorHandler, firmwareBinFile.size)
-            )
-            .andThen(
-                createFirmwareChunks(firmwareBinFile)
-                    .pairWithProgress()
-                    .toObservable()
-            )
-            .concatMap { (chunk, progress) ->
-                sendImageChunk(cypressOtaMessageChannel, errorHandler, chunk)
-                    .andThen(Observable.just(progress))
-            }
-            .concatWith(
-                sendVerifyImageCommand(cypressOtaMessageChannel, errorHandler, crc32Calculator.calculateCrc32(firmwareBinFile))
+    suspend fun program(
+        cypressOtaMessageChannel: CypressOtaMessageChannel, firmwareBinFile: ByteArray
+    ): Flow<Float> {
+        sendPrepareDownloadCommand(cypressOtaMessageChannel)
+        sendDownloadCommand(cypressOtaMessageChannel, firmwareBinFile.size)
+        val chunks = createFirmwareChunks(firmwareBinFile).pairWithProgress()
+        return chunks.asFlow().map { (chunk, progress) ->
+            sendImageChunk(cypressOtaMessageChannel, chunk)
+            progress
+        }.onCompletion {
+            //verify OTA is OK
+            sendVerifyImageCommand(
+                cypressOtaMessageChannel, crc32Calculator.calculateCrc32(firmwareBinFile)
             )
+        }
+    }
 
-    private fun sendPrepareDownloadCommand(cypressOtaMessageChannel: CypressOtaMessageChannel, errorHandler: ResponseErrorHandler): Completable =
-        sendCypressOtaModeCommandAndReceiveResponse<CypressOtaResponse>(cypressOtaMessageChannel, errorHandler,
-            PrepareDownloadCommand()
-        ).verifyResponseIs(CypressOtaResponseType.OK, "PrepareDownloadCommand")
+    private suspend fun sendPrepareDownloadCommand(
```

Comment:
Something like this would be a bit easier to read:

```
    private suspend fun sendPrepareDownloadCommand(
        cypressOtaMessageChannel: CypressOtaMessageChannel
    ) = cypressOtaMessageChannel
        .sendCypressOtaModeCommandAndReceiveResponse<CypressOtaResponse>(PrepareDownloadCommand())
        .verifyResponseIs(CypressOtaResponseType.OK, "PrepareDownloadCommand")

```